usd/src/sdf_parser/Joint.cc

```diff
@@ -0,0 +1,402 @@
+#include <pxr/usd/usd/stage.h>
+#include <pxr/usd/usdPhysics/driveAPI.h>
+#include <pxr/usd/usdPhysics/fixedJoint.h>
+#include <pxr/usd/usdPhysics/joint.h>
+#include <pxr/usd/usdPhysics/prismaticJoint.h>
+#include <pxr/usd/usdPhysics/revoluteJoint.h>
+#include <pxr/usd/usdPhysics/sphericalJoint.h>
+#pragma pop_macro ("__DEPRECATED")
+
+#include "sdf/Error.hh"
+#include "sdf/Joint.hh"
+#include "sdf/JointAxis.hh"
+#include "sdf/Link.hh"
+#include "sdf/Model.hh"
+#include "../UsdUtils.hh"
+
+namespace sdf
+{
+// Inline bracke to help doxygen filtering.
+inline namespace SDF_VERSION_NAMESPACE {
+//
+namespace usd
+{
+  /// \brief Helper function for setting a USD joint's pose relative to the
+  /// joint's parent and child.
+  /// \param[in] _jointPrim The USD joint prim
+  /// \param[in] _joint The SDF representation of _jointPrim
+  /// \param[in] _parentModel The SDF model that is the parent of _joint
+  /// \return UsdErrors, which is a vector of UsdError objects. Each UsdError
+  /// includes an error code and message. An empty vector indicates no error
+  /// occurred when setting _joint's pose relative to _joint's parent and child
+  UsdErrors SetUSDJointPose(pxr::UsdPhysicsJoint &_jointPrim,
+      const sdf::Joint &_joint, const sdf::Model &_parentModel)
+  {
+    UsdErrors errors;
+
+    ignition::math::Pose3d parentToJoint;
+    if (_joint.ParentLinkName() == "world")
+    {
+      ignition::math::Pose3d modelToJoint;
+      auto poseErrors = usd::PoseWrtParent(_joint, modelToJoint);
+      if (!poseErrors.empty())
+        return poseErrors;
+
+      // it is assumed the _parentModel's parent is the world
+      ignition::math::Pose3d worldToModel;
+      poseErrors = usd::PoseWrtParent(_parentModel, worldToModel);
+      if (!poseErrors.empty())
+        return poseErrors;
+
+      parentToJoint = worldToModel * modelToJoint;
+    }
+    else
+    {
+      auto poseResolutionErrors =
+        _joint.SemanticPose().Resolve(parentToJoint, _joint.ParentLinkName());
+      if (!poseResolutionErrors.empty())
+      {
+        errors.push_back(UsdError(
+              sdf::Error(sdf::ErrorCode::POSE_RELATIVE_TO_INVALID,
+              "Unable to get the pose of joint [" + _joint.Name() +
+              "] w.r.t. its parent link [" + _joint.ParentLinkName() + "].")));
+        for (const auto &e : poseResolutionErrors)
+          errors.push_back(UsdError(e));
+        return errors;
+      }
+    }
+
+    ignition::math::Pose3d childToJoint;
+    auto poseResolutionErrors = _joint.SemanticPose().Resolve(childToJoint,
+        _joint.ChildLinkName());
+    if (!poseResolutionErrors.empty())
+    {
+      errors.push_back(UsdError(
+          sdf::Error(sdf::ErrorCode::POSE_RELATIVE_TO_INVALID,
+            "Unable to get the pose of joint [" + _joint.Name() +
+            "] w.r.t. its child [" + _joint.ChildLinkName() + "].")));
+      for (const auto &e : poseResolutionErrors)
+        errors.push_back(UsdError(e));
```

Comment:
We will need to keep this loop since this case is different from other error cases; see https://github.com/ignitionrobotics/sdformat/pull/837#discussion_r814082594 for an explanation.